query.go

```diff
@@ -785,43 +790,127 @@ func (s *ChainService) prepareCFiltersQuery(blockHash chainhash.Hash,
+	case forwardBatch:
+		startHeight = int64(height)
+		stopHeight = startHeight + wire.MaxGetCFiltersReqRange - 1
+
+		// We need a longer timeout, since we are going to receive more
+		// than a single response.
+		options = append(options, Timeout(QueryBatchTimeout))
+
+	// Reverse batch, fetch as many of the preceding filters as possible.
+	case reverseBatch:
+		stopHeight = int64(height)
+		startHeight = stopHeight - wire.MaxGetCFiltersReqRange + 1
+
+		// We need a longer timeout, since we are going to receive more
+		// than a single response.
+		options = append(options, Timeout(QueryBatchTimeout))
 	}
 
-	log.Debugf("Fetching filter for height=%v, hash=%v", height, blockHash)
+	// Block 1 is the earliest one we can fetch.
+	if startHeight < 1 {
+		startHeight = 1
+	}
+
+	// If the stop height with the maximum batch size is above our best
+	// known block, then we use the best block height instead.
+	if stopHeight > bestHeight {
+		stopHeight = bestHeight
+	}
 
-	// In addition to fetching the block header, we'll fetch the filter
-	// headers (for this particular filter type) from the database. These
-	// are required in order to verify the authenticity of the filter.
-	curHeader, err := getHeader(&blockHash)
+	stopHash, err := s.GetBlockHash(stopHeight)
 	if err != nil {
-		return nil, fmt.Errorf("Couldn't get cfheader for block %s "+
-			"from database", blockHash)
+		return nil, fmt.Errorf("unable to get hash for "+
+			"stopHeight=%d: %v", stopHeight, err)
 	}
-	prevHeader, err := getHeader(&block.PrevBlock)
+
+	// In order to verify the authenticity of the received filters, we'll
+	// fetch the block headers and filter headers in the range
+	// [startHeight-1, stopHeight]. We go one below our startHeight since
+	// the hash of the previous block is needed for validation.
+	numFilters := uint32(stopHeight - startHeight + 1)
+	blockHeaders, _, err := s.BlockHeaders.FetchHeaderAncestors(
+		numFilters, stopHash,
+	)
 	if err != nil {
-		return nil, fmt.Errorf("Couldn't get cfheader for block %s "+
-			"from database", blockHash)
+		return nil, fmt.Errorf("unable to get %d blockheader "+
+			"ancestors for stopHash=%v: %v", numFilters,
+			stopHash, err)
+	}
+
+	if len(blockHeaders) != int(numFilters)+1 {
+		return nil, fmt.Errorf("expected %d block headers, got %d",
+			numFilters+1, len(blockHeaders))
+	}
+
+	filterHeaders, _, err := s.RegFilterHeaders.FetchHeaderAncestors(
+		numFilters, stopHash,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get %d filter header "+
+			"ancestors for stopHash=%v: %v", numFilters, stopHash,
+			err)
+	}
+
+	if len(filterHeaders) != int(numFilters)+1 {
+		return nil, fmt.Errorf("expected %d filter headers, got %d",
+			numFilters+1, len(filterHeaders))
 	}
 
+	// We create a header index such that we can easily index into our
+	// header slices for a given block hash in the received response,
+	// without consulting the database. This also keeps track of for which
```

Comment:
Nit: s/for//

Reply:
https://github.com/lightninglabs/neutrino/pull/134/commits/36d528abfa9d4c8ee4fa88f56340e77ba55bc1e0